myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java

```diff
@@ -37,103 +37,124 @@
-                            SchedulerState schedulerState,
-                            NMProfileManager profileManager,
+    public MyriadOperations(SchedulerState schedulerState,
                             NodeScaleDownPolicy nodeScaleDownPolicy) {
-        this.cfg = cfg;
-        this.schedulerState = schedulerState;
-        this.profileManager = profileManager;
-        this.nodeScaleDownPolicy = nodeScaleDownPolicy;
+      this.schedulerState = schedulerState;
+      this.nodeScaleDownPolicy = nodeScaleDownPolicy;
     }
 
-    public void flexUpCluster(int instances, String profile) {
+    public void flexUpCluster(NMProfile profile, int instances, Constraint constraint) {
         Collection<NodeTask> nodes = new HashSet<>();
         for (int i = 0; i < instances; i++) {
-            nodes.add(new NodeTask(profileManager.get(profile)));
+            nodes.add(new NodeTask(profile, constraint));
         }
 
-        LOGGER.info("Adding {} instances to cluster", nodes.size());
         this.schedulerState.addNodes(nodes);
     }
 
-    public void flexDownCluster(int numInstancesToScaleDown) {
-        LOGGER.info("About to flex down {} instances", numInstancesToScaleDown);
+    public void flexDownCluster(NMProfile profile, Constraint constraint, int numInstancesToScaleDown) {
+        // Flex down Pending tasks, if any
+        int numPendingTasksScaledDown = 0;
+          Set<Protos.TaskID> pendingTasks = Sets.newHashSet(this.schedulerState.getPendingTaskIds());
 
-        int numScaledDown = 0;
-        Set<NodeTask> activeTasks = Sets.newHashSet(this.schedulerState.getActiveTasks());
-        List<String> nodesToScaleDown = nodeScaleDownPolicy.getNodesToScaleDown();
-        if (activeTasks.size() > nodesToScaleDown.size()) {
-            LOGGER.info("Will skip flexing down {} Node Manager instances that were launched but " +
-                    "have not yet registered with Resource Manager.", activeTasks.size() - nodesToScaleDown.size());
-        }
-        
-        // If a NM is flexed down it takes time for the RM to realize the NM is no longer up
-        // We need to make sure we filter out nodes that have already been flexed down
-        // but have not disappeared from the RM's view of the cluster
-        for (Iterator<String> iterator = nodesToScaleDown.iterator(); iterator.hasNext();) {
-            String nodeToScaleDown = iterator.next();
-            boolean nodePresentInMyriad = false;
-            for (NodeTask nodeTask : activeTasks) {
-                if (nodeTask.getHostname().equals(nodeToScaleDown)) {
-                    nodePresentInMyriad = true;    
-                    break;
-                }
-            }
-            if (!nodePresentInMyriad) {
-                iterator.remove();
+          for (Protos.TaskID taskId : pendingTasks) {
+            NodeTask nodeTask = schedulerState.getTask(taskId);
+            if (nodeTask.getProfile().getName().equals(profile.getName()) &&
```

Comment:
nodeTask potentially could be null, as we get set of pendingTasks get prior and it can be removed meanwhile.